**Calling the packages required for the work**

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 21.7 MB/s 
     |████████████████████████████████| 49 kB 5.8 MB/s 


In [2]:
from pyomo.environ import * 

In [3]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155673 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

$\large{\textbf{Solution to Exercise 1:}}$

In [4]:
# Create a model
model1 = ConcreteModel()

In [5]:
#Declare decision variables
model1.x1 = Var(domain=Reals)
model1.x2 = Var(domain=NonNegativeReals)
model1.x3 = Var(domain=Reals)
model1.x4 = Var(domain=Reals)
model1.x5 = Var(domain=Reals)
model1.x6 = Var(domain=NonNegativeReals)
model1.x7 = Var(domain=Reals)

In [6]:
#Declare objective function
model1.obj1 = Objective(expr = -3*model1.x1 - 2*model1.x2 + 8*model1.x3 - 7*model1.x4 + 0.19*model1.x5 + 11*model1.x6 + 20.2*model1.x7, sense=minimize)

In [7]:
#Declare constraints
model1.cons1 = Constraint(expr = model1.x1 + model1.x2 + model1.x3 + model1.x4 + model1.x5 + model1.x6 + model1.x7 == 256)
model1.cons2 = Constraint(expr = -model1.x2 + model1.x3 >=10)
model1.cons3 = Constraint(expr = -3*model1.x1 + model1.x5 - 12*model1.x6 <= 105)
model1.cons4 = Constraint(expr = 18.5*model1.x2 + 30.5*model1.x4 - model1.x7 >= 28)
model1.cons5 = Constraint(expr = -model1.x3 - 4*model1.x5 + model1.x7 >= 31)
model1.cons6 = Constraint(expr = model1.x1>= -5)
model1.cons7 = Constraint(expr = model1.x3>= 5)
model1.cons8 = Constraint(expr = model1.x4>= -3)
model1.cons9 = Constraint(expr = model1.x5>= -4)
model1.cons10 = Constraint(expr = model1.x7>= 10)

In [8]:
model1.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x7 : Size=1, Index=None
        Key  : Lower 

In [9]:
SolverFactory('glpk', executable= '/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1010.76
  Upper bound: -1010.76
  Number of objectives: 1
  Number of constraints: 11
  Number of variables: 8
  Number of nonzeros: 24
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.02539825439453125
# ----------------------------------------------------------
#   Solution Information
# -----------------------------

In [10]:
#Displaying the optimal solution
print('\nCost = ', model1.obj1())

print('\nDecision Variables')
print('x1 = ', model1.x1())
print('x2 = ', model1.x2())
print('x3 = ', model1.x3())
print('x4 = ', model1.x4())
print('x5 = ', model1.x5())
print('x6 = ', model1.x6())
print('x7 = ', model1.x7())

print('\nConstraints')
print('Constraint 1  = ', model1.cons1())
print('Constraint 2  = ', model1.cons2())
print('Constraint 3  = ', model1.cons3())
print('Constraint 4  = ', model1.cons4())
print('Constraint 5  = ', model1.cons5())


Cost =  -1010.760000000002

Decision Variables
x1 =  -5.0
x2 =  0.0
x3 =  10.0
x4 =  230.0
x5 =  -4.0
x6 =  0.0
x7 =  24.9999999999999

Constraints
Constraint 1  =  255.9999999999999
Constraint 2  =  10.0
Constraint 3  =  11.0
Constraint 4  =  6990.0
Constraint 5  =  30.9999999999999


$\large{\mathbf{Observations:}}$

$\mathbf{1.3}$

Objective Function Value = -1010.76

$x_1 = -5 \\ x_2 = 0 \\ x_3 = 10 \\ x_4 = 230 \\ x_5 = -4 \\ x_6 = 0 \\ x_7 = 25$

$\mathbf{1.4} \\ \textbf{Constraint 1:} \ \text{It is active as value of LHS = 256 which satisfies the equality constraint.} \\ \textbf{Constraint 2:} \ \text{It is active as value of LHS = 10 which satisfies the equality constraint.} \\ \textbf{Constraint 3:} \text{ It is inactive as value if LHS = 11 which is strictly less than RHS = 105.} \\ \textbf{Constraint 4:} \text{ It is inactive as value of LHS = 6990 which is strictly greater than RHS = 28.} \\ \textbf{Constraint 5:} \text{ It is active as value of LHS = 31 which satisfies the equality constraint.}$

$\large{\textbf{Solution to Exercise 2:}}$

$\large{\mathbf{2.1 \ Formulation:}}$


$\\ \\ \large{\text{Assumption 1:}}\ \\ -x_2 - x_4 + x_5 \geq ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \geq \epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.} $

$\textbf{Problem:} \\ \text{max } 0.082x_1 + 0.036x_2 + 0.0156x_3 + 0.033x_4 + 0.054x_5 + 0.0125y \\ \text{s.t. } 3x_1 - 3x_2 + 5x_3 - x_4 + x_5 - y \leq 80 \\ x_3 + 2x_4 + 4y \geq 30 \\ 3x_1 + 3.6x_2 - 1.5x_3 + 5.6x_4 - 1.8x_5 + 6x_6 \leq 0 \\ 7x_1 - 2x_2 - 3x_3 + 4x_4 + 5x_5 - 6x_6 + 12y \leq 0 \\ x_1,x_2,x_3,x_4,x_5,x_6 \geq 0 \\ -0.2 \leq y \leq 5.5$



In [11]:
!pip install -q pyomo

In [12]:
from pyomo.environ import * 

In [13]:
!apt-get install -y -qq glpk-utils

In [14]:
# Create a model
model2 = ConcreteModel()

In [15]:
#Declare decision variables
model2.x1 = Var(domain=NonNegativeReals)
model2.x2 = Var(domain=NonNegativeReals)
model2.x3 = Var(domain=NonNegativeReals)
model2.x4 = Var(domain=NonNegativeReals)
model2.x5 = Var(domain=NonNegativeReals)
model2.x6 = Var(domain=NonNegativeReals)
model2.y = Var(domain=Reals)

In [16]:
#Declare objective function
model2.obj2 = Objective(expr = 0.082*model2.x1 + 0.036*model2.x2 + 0.0156*model2.x3 + 0.033*model2.x4 + 0.054*model2.x5 + 0.0125*model2.y, sense=maximize)

In [17]:
#Declare constraints
model2.cons1 = Constraint(expr = 3*model2.x1 - 3*model2.x2 + 5*model2.x3 - model2.x4 + model2.x5 - model2.y <= 80)
model2.cons2 = Constraint(expr = model2.x3 + 2*model2.x4 + 4*model2.y >= 30)
model2.cons3 = Constraint(expr = 3*model2.x1 + 3.6*model2.x2 - 1.5*model2.x3 + 5.6*model2.x4 - 1.8*model2.x5 + 6*model2.x6 <= 0)
model2.cons4 = Constraint(expr = 7*model2.x1 - 2*model2.x2 - 3*model2.x3 + 4*model2.x4 + 5*model2.x5 - 6*model2.x6 + 12*model2.y <= 0)
model2.cons5 = Constraint(expr = model2.y >= -0.2)
model2.cons6 = Constraint(expr = model2.y <= 5.5)

In [18]:
model2.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    y : S

In [19]:
SolverFactory('glpk', executable= '/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 2.12102483443709
  Upper bound: 2.12102483443709
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 8
  Number of nonzeros: 25
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.04104757308959961
# ----------------------------------------------------------
#   Solution Information
# --------------

In [20]:
print('\nCost = ', model2.obj2())

print('\nDecision Variables')
print('x1 = ', model2.x1())
print('x2 = ', model2.x2())
print('x3 = ', model2.x3())
print('x4 = ', model2.x4())
print('x5 = ', model2.x5())
print('x6 = ', model2.x6())
print('y = ', model2.y())

print('\nConstraints')
print('Constraint 1  = ', model2.cons1())
print('Constraint 2  = ', model2.cons2())
print('Constraint 3  = ', model2.cons3())
print('Constraint 4  = ', model2.cons4())


Cost =  2.1210248344370837

Decision Variables
x1 =  0.0
x2 =  19.5529801324503
x3 =  24.2384105960265
x4 =  0.0
x5 =  18.9072847682119
x6 =  0.0
y =  1.44039735099338

Constraints
Constraint 1  =  80.00000000000011
Constraint 2  =  30.00000000000002
Constraint 3  =  -9.237055564881302e-14
Constraint 4  =  -6.039613253960852e-14


In [21]:
print(-model2.x2() - model2.x4() + model2.x5())
print(-model2.x1() + model2.x3() + model2.x6() - 2*model2.y())

-0.6456953642384029
21.35761589403974


$\large{\textbf{Observations:}} \\ -x_2 - x_4 + x_5 = -19.55 - 24.238 + 18.907 = -0.64569  \\ \text{ is coming out to be negative which is a contradiction to our assumption.} \\ \text{Hence there is no optimal solution to the problem when} \\ -x_2 - x_4 + x_5 \geq ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \geq \epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.}$

$\\ \\ \large{\textbf{Assumption 2:}}\ \\ -x_2 - x_4 + x_5 \geq ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \leq -\epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.} $

$\textbf{Problem:} \\ \text{max } 0.082x_1 + 0.036x_2 + 0.0156x_3 + 0.033x_4 + 0.054x_5 + 0.0125y \\ \text{s.t. } 3x_1 - 3x_2 + 5x_3 - x_4 + x_5 - y \leq 80 \\ x_3 + 2x_4 + 4y \geq 30 \\ 3x_1 + 3.6x_2 - 1.5x_3 + 5.6x_4 - 1.8x_5 + 6x_6 \leq 0 \\ 7x_1 - 2x_2 - 3x_3 + 4x_4 + 5x_5 - 6x_6 + 12y \geq 0 \\ x_1,x_2,x_3,x_4,x_5,x_6 \geq 0 \\ -0.2 \leq y \leq 5.5$


In [22]:
!pip install -q pyomo

In [23]:
from pyomo.environ import * 

In [24]:
!apt-get install -y -qq glpk-utils

In [25]:
# Create a model
model22 = ConcreteModel()

In [26]:
#Declare decision variables
model22.x1 = Var(domain=NonNegativeReals)
model22.x2 = Var(domain=NonNegativeReals)
model22.x3 = Var(domain=NonNegativeReals)
model22.x4 = Var(domain=NonNegativeReals)
model22.x5 = Var(domain=NonNegativeReals)
model22.x6 = Var(domain=NonNegativeReals)
model22.y = Var(domain=Reals)

In [27]:
#Declare objective function
model22.obj2 = Objective(expr = 0.082*model22.x1 + 0.036*model22.x2 + 0.0156*model22.x3 + 0.033*model22.x4 + 0.054*model22.x5 + 0.0125*model22.y, sense=maximize)

In [28]:
#Declare constraints
model22.cons1 = Constraint(expr = 3*model22.x1 - 3*model22.x2 + 5*model22.x3 - model22.x4 + model22.x5 - model22.y <= 80)
model22.cons2 = Constraint(expr = model22.x3 + 2*model22.x4 + 4*model22.y >= 30)
model22.cons3 = Constraint(expr = 3*model22.x1 + 3.6*model22.x2 - 1.5*model22.x3 + 5.6*model22.x4 - 1.8*model22.x5 + 6*model22.x6 <= 0)
model22.cons4 = Constraint(expr = 7*model22.x1 - 2*model22.x2 - 3*model22.x3 + 4*model22.x4 + 5*model22.x5 - 6*model22.x6 + 12*model22.y >= 0)
model22.cons5 = Constraint(expr = model22.y >= -0.2)
model22.cons6 = Constraint(expr = model22.y <= 5.5)

In [29]:
model22.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    y : S

In [30]:
SolverFactory('glpk', executable= '/usr/bin/glpsol').solve(model22).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 8
  Number of nonzeros: 25
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.05690360069274902


$\large{\textbf{Observations:}}$

$\text{We can observe that no optimal solution exists in this case as the lower bound is coming out to be } -∞ \text{ while the upper bound is } ∞. \\ \text{Hence there is no optimal solution to the problem when} \\ -x_2 - x_4 + x_5 \geq ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \leq -\epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.}$

$\\ \\ \large{\textbf{Assumption 3:}}\ \\ -x_2 - x_4 + x_5 \leq -ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \geq \epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.} $

$\textbf{Problem:} \\ \text{max } 0.082x_1 + 0.036x_2 + 0.0156x_3 + 0.033x_4 + 0.054x_5 + 0.0125y \\ \text{s.t. } 3x_1 - 3x_2 + 5x_3 - x_4 + x_5 - y \leq 80 \\ x_3 + 2x_4 + 4y \geq 30 \\ 3x_1 + 3.6x_2 - 1.5x_3 + 5.6x_4 - 1.8x_5 + 6x_6 \geq 0 \\ 7x_1 - 2x_2 - 3x_3 + 4x_4 + 5x_5 - 6x_6 + 12y \leq 0 \\ x_1,x_2,x_3,x_4,x_5,x_6 \geq 0 \\ -0.2 \leq y \leq 5.5$

In [31]:
!pip install -q pyomo

In [32]:
from pyomo.environ import *

In [33]:
!apt-get install -y -qq glpk-utils

In [34]:
# Create a model
model23 = ConcreteModel()

In [35]:
#Declare decision variables
model23.x1 = Var(domain=NonNegativeReals)
model23.x2 = Var(domain=NonNegativeReals)
model23.x3 = Var(domain=NonNegativeReals)
model23.x4 = Var(domain=NonNegativeReals)
model23.x5 = Var(domain=NonNegativeReals)
model23.x6 = Var(domain=NonNegativeReals)
model23.y = Var(domain=Reals)

In [36]:
#Declare objective function
model23.obj2 = Objective(expr = 0.082*model23.x1 + 0.036*model23.x2 + 0.0156*model23.x3 + 0.033*model23.x4 + 0.054*model23.x5 + 0.0125*model23.y, sense=maximize)

In [37]:
#Declare constraints
model23.cons1 = Constraint(expr = 3*model23.x1 - 3*model23.x2 + 5*model23.x3 - model23.x4 + model23.x5 - model23.y <= 80)
model23.cons2 = Constraint(expr = model23.x3 + 2*model23.x4 + 4*model23.y >= 30)
model23.cons3 = Constraint(expr = 3*model23.x1 + 3.6*model23.x2 - 1.5*model23.x3 + 5.6*model23.x4 - 1.8*model23.x5 + 6*model23.x6 >= 0)
model23.cons4 = Constraint(expr = 7*model23.x1 - 2*model23.x2 - 3*model23.x3 + 4*model23.x4 + 5*model23.x5 - 6*model23.x6 + 12*model23.y <= 0)
model23.cons5 = Constraint(expr = model23.y >= -0.2)
model23.cons6 = Constraint(expr = model23.y <= 5.5)

In [38]:
model23.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    y : S

In [39]:
SolverFactory('glpk', executable= '/usr/bin/glpsol').solve(model23).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 8
  Number of nonzeros: 25
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.05254364013671875


$\large{\textbf{Observations:}}$

$\text{We can observe that no optimal solution exists in this case as the lower bound is coming out to be } -∞ \text{ while the upper bound is } ∞. \\ \text{Hence there is no optimal solution to the problem when} \\ -x_2 - x_4 + x_5 \leq -ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \geq \epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.}$

$\\ \\ \large{\textbf{Assumption 4:}}\ \\ -x_2 - x_4 + x_5 \leq -ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \leq -\epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.} $

$\textbf{Problem:} \\ \text{max } 0.082x_1 + 0.036x_2 + 0.0156x_3 + 0.033x_4 + 0.054x_5 + 0.0125y \\ \text{s.t. } 3x_1 - 3x_2 + 5x_3 - x_4 + x_5 - y \leq 80 \\ x_3 + 2x_4 + 4y \geq 30 \\ 3x_1 + 3.6x_2 - 1.5x_3 + 5.6x_4 - 1.8x_5 + 6x_6 \geq 0 \\ 7x_1 - 2x_2 - 3x_3 + 4x_4 + 5x_5 - 6x_6 + 12y \geq 0 \\ x_1,x_2,x_3,x_4,x_5,x_6 \geq 0 \\ -0.2 \leq y \leq 5.5$

In [40]:
!pip install -q pyomo

In [41]:
from pyomo.environ import *

In [42]:
!apt-get install -y -qq glpk-utils

In [43]:
# Create a model
model24 = ConcreteModel()

In [44]:
#Declare decision variables
model24.x1 = Var(domain=NonNegativeReals)
model24.x2 = Var(domain=NonNegativeReals)
model24.x3 = Var(domain=NonNegativeReals)
model24.x4 = Var(domain=NonNegativeReals)
model24.x5 = Var(domain=NonNegativeReals)
model24.x6 = Var(domain=NonNegativeReals)
model24.y = Var(domain=Reals)

In [45]:
#Declare objective function
model24.obj2 = Objective(expr = 0.082*model24.x1 + 0.036*model24.x2 + 0.0156*model24.x3 + 0.033*model24.x4 + 0.054*model24.x5 + 0.0125*model24.y, sense=maximize)

In [46]:
#Declare constraints
model24.cons1 = Constraint(expr = 3*model24.x1 - 3*model24.x2 + 5*model24.x3 - model24.x4 + model24.x5 - model24.y <= 80)
model24.cons2 = Constraint(expr = model24.x3 + 2*model24.x4 + 4*model24.y >= 30)
model24.cons3 = Constraint(expr = 3*model24.x1 + 3.6*model24.x2 - 1.5*model24.x3 + 5.6*model24.x4 - 1.8*model24.x5 + 6*model24.x6 >= 0)
model24.cons4 = Constraint(expr = 7*model24.x1 - 2*model24.x2 - 3*model24.x3 + 4*model24.x4 + 5*model24.x5 - 6*model24.x6 + 12*model24.y >= 0)
model24.cons5 = Constraint(expr = model24.y >= -0.2)
model24.cons6 = Constraint(expr = model24.y <= 5.5)

In [47]:
model24.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    y : S

In [48]:
SolverFactory('glpk', executable= '/usr/bin/glpsol').solve(model24).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 8
  Number of nonzeros: 25
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.03890490531921387


$\large{\textbf{Observations:}}$

$\text{We can observe that no optimal solution exists in this case as the lower bound is coming out to be } -∞ \text{ while the upper bound is } ∞. \\ \text{Hence there is no optimal solution to the problem when} \\ -x_2 - x_4 + x_5 \leq -ϵ_1 \ \\ -x_1 + x_3 + x_6 - 2y \leq -\epsilon_2 \\ \text{where} \ ϵ_1, ϵ_2 \text{ are very small positive reals.}$

$\large{\textbf{Solution to Exercise 3:}}$

$\large{\mathbf{3.1 \ Formulation:}}$

$\min \ 1200x_1 + 2580x_2 \ \\ \text{s.t.} \ x_1 + x_2 \geq 145\  \text{(Quantity Constraint)} \\ 2.18x_1 + 1.083x_2 \leq 980 \ \text{(Time Constraint)}\\ x_1 \geq 105 \ \text{(Minimum Order)} \\ x_2 \geq 125 \ \text{(Minimum Order)}$

$\text{where} \ x_1 \ \text{is the quantity of P produced and} \ x_2 \  \text{is the quantity of Q produced. }$

In [49]:
!pip install -q pyomo

In [50]:
from pyomo.environ import *

In [51]:
#Create a model
model3 = ConcreteModel()

In [52]:
#Declare decision variables
model3.x1 = Var(domain=Reals)
model3.x2 = Var(domain=Reals)

In [53]:
#Declare objective function
model3.obj3 = Objective(expr = 1200*model3.x1 + 2580*model3.x2, sense=minimize)

In [54]:
#Declare contraints
model3.const1 = Constraint(expr = model3.x1 + model3.x2 >= 145)
model3.const2 = Constraint(expr = 2.18*model3.x1 + 1.083*model3.x2 <= 980)
model3.const3 = Constraint(expr = model3.x1 >= 105)
model3.const4 = Constraint(expr = model3.x2 >= 125)

In [55]:
#Print the model
model3.pprint()

2 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj3 : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 1200*x1 + 2580*x2

4 Constraint Declarations
    const1 : Size=1, Index=None, Active=True
        Key  : Lower : Body    : Upper : Active
        None : 145.0 : x1 + x2 :  +Inf :   True
    const2 : Size=1, Index=None, Active=True
        Key  : Lower : Body               : Upper : Active
        None :  -Inf : 2.18*x1 + 1.083*x2 : 980.0 :   True
    const3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None : 105.0 :   x1 :  +Inf :   True
    const4 : Size=1, Index=None, Active=Tru

In [56]:
!apt-get install -y -qq glpk-utils

In [57]:
SolverFactory('glpk',executable='/usr/bin/glpsol').solve(model3).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 448500.0
  Upper bound: 448500.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.02421855926513672
# ----------------------------------------------------------
#   Solution Information
# -------------------------------

In [58]:
#Print solutions
print('\nCost = ', model3.obj3())

print('\nDecision Variables')
print('x1 = ', model3.x1())
print('x2 = ', model3.x2())

print('\nConstraints')
print('Constraint 1  = ', model3.const1())
print('Constraint 2  = ', model3.const2())
print('Constraint 3  = ', model3.const3())
print('Constraint 4  = ', model3.const4())


Cost =  448500.0

Decision Variables
x1 =  105.0
x2 =  125.0

Constraints
Constraint 1  =  230.0
Constraint 2  =  364.275
Constraint 3  =  105.0
Constraint 4  =  125.0


$\mathbf{\large{Observations:}} \\ \mathbf{3.3} \\ \text{Objective Function Value} = 448500 \\ \textbf{Decision Variables:} \\ x_1 = 105 \\ x_2 = 125 \\ \textbf{Constraint 1:} \text{ It is inactive as value of LHS = 230 is strictly greater than RHS = 145.} \\ \textbf{Constraint 2:} \text{ It is inactive as value of LHS = 364.275 is strictly less than RHS = 980} \\ \textbf{Constraint 3:} \text{ It is active as value of LHS = 105 which satisfies the equality constraint.} \\ \textbf{Constraint 4:} \text{ It is active as value of LHS = 125 which satisfies the equality constraint.}$

$\large{\mathbf{3.4 \ New \ Formulation:}}$

$\min \ 1200x_1 + 2580x_2 \ \\ \text{s.t.} \ x_1 + x_2 \geq 145\  \text{(Quantity Constraint)} \\ 2.18x_1 + 1.083x_2 \leq 980 \ \text{(Time Constraint)}\\ x_1 \geq 147 \ \text{(Additional Order)} \\ x_2 \geq 222 \ \text{(Additional Order)}$

$\text{where} \ x_1 \ \text{is the quantity of P produced and} \ x_2 \  \text{is the quantity of Q produced. }$

In [59]:
!pip install -q pyomo

In [60]:
from pyomo.environ import *

In [61]:
#Create a model
model4 = ConcreteModel()

In [62]:
#Declare decision variables
model4.x1 = Var(domain=Reals)
model4.x2 = Var(domain=Reals)

In [63]:
#Declare objective function
model4.obj4 = Objective(expr = 1200*model4.x1 + 2580*model4.x2, sense=minimize)

In [64]:
#Declare contraints
model4.const1 = Constraint(expr = model4.x1 + model4.x2 >= 145)
model4.const2 = Constraint(expr = 2.18*model4.x1 + 1.083*model4.x2 <= 980)
model4.const3 = Constraint(expr = model4.x1 >= 147)
model4.const4 = Constraint(expr = model4.x2 >= 222)

In [65]:
#Print the model
model4.pprint()

2 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj4 : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 1200*x1 + 2580*x2

4 Constraint Declarations
    const1 : Size=1, Index=None, Active=True
        Key  : Lower : Body    : Upper : Active
        None : 145.0 : x1 + x2 :  +Inf :   True
    const2 : Size=1, Index=None, Active=True
        Key  : Lower : Body               : Upper : Active
        None :  -Inf : 2.18*x1 + 1.083*x2 : 980.0 :   True
    const3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None : 147.0 :   x1 :  +Inf :   True
    const4 : Size=1, Index=None, Active=Tru

In [66]:
!apt-get install -y -qq glpk-utils

In [67]:
SolverFactory('glpk',executable='/usr/bin/glpsol').solve(model4).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 749160.0
  Upper bound: 749160.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.0178985595703125
# ----------------------------------------------------------
#   Solution Information
# --------------------------------

In [68]:
#Print solutions
print('\nCost = ', model4.obj4())

print('\nDecision Variables')
print('x1 = ', model4.x1())
print('x2 = ', model4.x2())

print('\nConstraints')
print('Constraint 1  = ', model4.const1())
print('Constraint 2  = ', model4.const2())
print('Constraint 3  = ', model4.const3())
print('Constraint 4  = ', model4.const4())


Cost =  749160.0

Decision Variables
x1 =  147.0
x2 =  222.0

Constraints
Constraint 1  =  369.0
Constraint 2  =  560.886
Constraint 3  =  147.0
Constraint 4  =  222.0


$\mathbf{\large{Observations:}} \\ \mathbf{3.6} \\ \text{Objective Function Value} = 749160 \\ \textbf{Decision Variables:} \\ x_1 = 147 \\ x_2 = 222 \\ \textbf{Constraint 1:} \text{ It is inactive as value of LHS = 369 is strictly greater than RHS = 145.} \\ \textbf{Constraint 2:} \text{ It is inactive as value of LHS = 560.886 is strictly less than RHS = 980} \\ \textbf{Constraint 3:} \text{ It is active as value of LHS = 147 which satisfies the equality constraint.} \\ \textbf{Constraint 4:} \text{ It is active as value of LHS = 222 which satisfies the equality constraint.} \\ \\  \text{Jen Industries should increase their production by the additional quantity ordered.}$